Create page: Set up Spack for building ACCESS models on Gadi (#789)
ACCESS-NRI/access-hive.org.au: docs/getting_started/spack.md

> [spack_environments]: https://spack.readthedocs.io/en/latest/environments.html

Reviewer: Why define `spack_environments`?

Author: Before it was used twice, but in the new version there's no real need. 
I'll change it to normal link without defining it.

# Set up Spack for building ACCESS models

!!! danger
    This page is tailored to experienced users and collaborators developing ACCESS models.

[Spack](https://spack.io/about/) is a build-from-source package manager, specifically designed to simplify the installation of scientific software on supercomputers.

> To use _Spack_, please familiarise yourself with the [Basic Usage instructions](https://spack.readthedocs.io/en/latest/basic_usage.html) and [Environments][spack_environments].

Reviewer: What was the benefit of not putting the URL directly  into the text?

Author: Same as above. 
Initially that url was used twice. Fixed now.

> We also reccommend that you refer to the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).

Reviewer: ```suggestion
We also recommend that you refer to the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).
```

## Prerequisites
These instructions are tailored specifically for _Gadi_. To use _Spack_ on _Gadi_, you must have an NCI account. For instructions on how to set up an account, refer to [Set Up your NCI Account](/getting_started/set_up_nci_account).

## Set up Spack on Gadi

### Create a directory for Spack

Create a directory on the filesystem where _Spack_ will be installed (e.g. `/g/data/$PROJECT/$USER/<myspack>`). This directory shall be referred to as `<myspack>`.

```

> mkdir <myspack> && cd $_

Reviewer: Keep it simple:
```suggestion
mkdir <myspack>
cd <myspack>
```

```

### Clone the relevant git repositories

```
git clone -c feature.manyFiles=true https://github.com/ACCESS-NRI/spack.git --branch releases/v0.22 --single-branch --depth=1
git clone https://github.com/ACCESS-NRI/spack-packages.git --branch main
git clone https://github.com/ACCESS-NRI/spack-config.git --branch main
```

### Link Spack configuration files to the Spack instance

```
ln -s -r -v spack-config/v0.22/gadi/* spack/etc/spack/
```

!!! success
    Your _Spack_ setup is complete!

## Test Spack (OPTIONAL)

To test _Spack_ we will create an [ACCESS-OM2](/models/configurations/access-om#access-om2) environment and build the relevant packages. Then, we will uninstall all the packages and remove the environment.

```
module purge

> cs <myspack>

Reviewer: ```suggestion
cd <myspack>
```

> # Set up Spack for model builds

Reviewer: # Set-up Spack for building models on Gadi

Author: 1. I would not write "Set-up" with a hyphen ("set up" is already present in different parts of the website).
2. Despite being a setup specific to _Gadi_, I would not specify it in the title. Most of the Hive Docs content is related specifically to _Gadi_, but we don't specify it all the time (also to avoid making the title too long).
In the prerequisites we mention what is needed to use _Spack_ on _Gadi_, which I think is enough for a user.

What about we change this to: 
"Set up Spack for building models"

Reviewer: That's fine.

Author: Do you think it makes sense to add "ACCESS" as in "Set up Spack for building ACCESS models"?

Reviewer: Good idea!

[Spack](https://spack.io/about/) is a build-from-source package manager, specifically designed to simplify the installation of scientific software on supercomputers.

To know more about _Spack_ usage, please familiarise yourself with the [Basic Usage instructions](https://spack.readthedocs.io/en/latest/basic_usage.html) and [Environments](https://spack.readthedocs.io/en/latest/environments.html).<br>

> We also reccommend you check the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).
> 
> 
> ## Prerequisites
> To use _Spack_ on _Gadi_, you must have an NCI account.<br>
> For instructions on how to set up an account, refer to [Set Up your NCI Account](/getting_started/set_up_nci_account).
> 
> ## Set up Spack on Gadi
> 
> ### Create a directory for the Spack distribution
> 
> !!! tip
>     This step is optional, but it is recommended you have a selected directory where to keep your _Spack_ distribution.
> 
> The complete _Spack_ distribution is approximately 120 MB in size. For this reason, we reccommend placing its directory somewhere in `/g/data`. An example can be `/g/data/$PROJECT/$USER/spack_distribution`.
> 
> To create the directory mentioned above and navigate into it, run the following command:
> ```
> mkdir -p /g/data/$PROJECT/$USER/spack_distribution && cd $_
> ```
> 
> ### Get Spack executable and configuration files
> 
> To get the _Spack_ executable and configuration files, run the following commands to clone the respective GitHub repositories:
> 
> ```
> git clone -c feature.manyFiles=true https://github.com/ACCESS-NRI/spack.git --branch releases/v0.22 --single-branch --depth=1
> git clone https://github.com/ACCESS-NRI/spack-packages.git --branch main
> git clone https://github.com/ACCESS-NRI/spack-config.git --branch main
> ```
> 
> ### Link Spack configuration files
> 
> To link the _Spack_ configuration files to the _Spack_ instance, run the following command:
> 
> ```
> ln -s -r -v spack-config/v0.22/gadi/* spack/etc/spack/
> ```
> 
> !!! success
>     Your _Spack_ setup is complete!
> 
> ### Test Spack (OPTIONAL)
> 
> To test _Spack_ we will create an environment (in this example it will be the on to build ACCESS-OM2 executables) and build the relevant packages. Then, we will uninstall all the packages and remove the environment.
> 
> All the steps listed above can be performed by running the following commands (where `<spack_distribution_directory>` is the [directory for the Spack distribution](#create-a-directory-for-the-spack-distribution)):
> ```
> . <spack_distribution_directory>/spack-config/spack-enable.bash
> curl -L https://raw.githubusercontent.com/ACCESS-NRI/ACCESS-OM2/main/spack.yaml -o access-om2-spack.yaml
> spack env create access-om2 access-om2-spack.yaml
> spack env activate -p access-om2
> spack find
> spack install --verbose
> spack find
> spack uninstall --remove --all -y
> spack env deactivate
> spack env rm access-om2 -y
> rm access-om2-spack.yaml
> ```

Reviewer: ```suggestion
# Set up Spack for building ACCESS models

[Spack](https://spack.io/about/) is a build-from-source package manager, specifically designed to simplify the installation of scientific software on supercomputers.

To use _Spack_, please familiarise yourself with the [Basic Usage instructions](https://spack.readthedocs.io/en/latest/basic_usage.html) and [Environments](https://spack.readthedocs.io/en/latest/environments.html).

We also recommend that you refer to the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).

## Prerequisites
These instructions are tailored specifically for _Gadi_. To use _Spack_ on _Gadi_, you must have an NCI account. For instructions on how to set up an account, refer to [Set Up your NCI Account](/getting_started/set_up_nci_account).

Create a directory on the filesystem where _Spack_ will be installed. This directory shall be refered to as `myspack`. e.g. `/g/data/$PROJECT/$USER/myspack`

## The steps for setting up Spack on Gadi are:

1. Change to the `myspack` directory:
```
cd myspack
```

2. Clone the relevant git repositories:
```
git clone -c feature.manyFiles=true https://github.com/ACCESS-NRI/spack.git --branch releases/v0.22 --single-branch --depth=1
git clone https://github.com/ACCESS-NRI/spack-packages.git --branch main
git clone https://github.com/ACCESS-NRI/spack-config.git --branch main
```

3. Link `spack-config` configurations files to the Spack instance:
```
ln -s -r -v spack-config/v0.22/gadi/* spack/etc/spack/
```

## Test Spack (OPTIONAL)

To test _Spack_ we will create an ACCESS-OM2 environment and build the relevant packages. Then, we will uninstall all the packages and remove the environment.

```
cd myspack
. spack-config/spack-enable.bash
git clone https://github.com/ACCESS-NRI/ACCESS-OM2.git
spack env create access-om2 ACCESS-OM2/spack.yaml
spack env activate access-om2
spack find
spack concretize -f
spack install --verbose
spack find
spack uninstall --remove --all
spack env deactivate
spack env rm access-om2
rm -rf ACCESS-OM2
```
```

> We also reccommend you check the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).

Reviewer: Move to the introduction because it is important information.
